implement system2 interfaces on top of descriptors

sigma opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

This change makes devnet-sdk's existing devnet descriptor a valid provide of system2 interfaces.
As a prerequisite, it bubbles up the interop  depset information through the descriptor.

<!--
A clear and concise description of the features you're adding in this pull request.
-->

**Tests**

<!--
Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.
-->

**Additional context**

<!--
Add any other context about the problem you're solving.
-->

**Metadata**

<!-- 
Include a link to any github issues that this may close in the following form:
- Fixes #[Link to Issue]
-->

sigma committed:
feat(kurtosis-devnet): expose interop depset

This change ensures that we are able to bubble up the dependency set
that kurtosis exposes since
https://github.com/ethpandaops/optimism-package/pull/205

sigma committed:
feat(devnet-sdk): new implementation of system2 interfaces

This maps devnet-sdk's descriptor to system2, effectively making
kurtosis and persitent devnets controllable through system2.

protolambda commented on devnet-sdk/systemext/l1.go:
```diff
@@ -0,0 +1,64 @@
+			ids := nodes[idx]
+
+			elRPC, err := findProtocolService(setup, ELServiceName, RPCProtocol, node.Services)
+			require.NoError(setup.T, err)
```
```suggestion
			setup.Require.NoError(err)
```
Nit (same with all other require usage).

protolambda commented on devnet-sdk/systemext/l2.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func defineSystemKeys(setup *system2.Setup) system2.L2Keys {
+	// TODO: get actual mnemonic from Kurtosis
```
```suggestion
	// TODO(#15040): improve key-loading from Kurtosis
```

Using the mnemonic is fine for now, but not what I'd like to do long-term. I opened an issue to track this work: https://github.com/ethereum-optimism/optimism/issues/15040

protolambda reviewed (dismissed):
Looks good, just a few nits

protolambda committed:
devnet-sdk: remove option footgun

protolambda committed:
devnet-sdk/systemext: review-fixes

protolambda committed:
devnet-sdk/system2: fix godoc wording

protolambda reviewed (commented):
Removed a footgun I noticed in the option code while trying this out. Removed `Append` since you had to remember to update the existing option reference. Instead we now have `Add`, which updates the existing reference immediately.

protolambda commented:
Trying to test with:

```go
package systemext

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/ethereum/go-ethereum/log"

	"github.com/ethereum-optimism/optimism/devnet-sdk/shell/env"
	"github.com/ethereum-optimism/optimism/devnet-sdk/system2"
	"github.com/ethereum-optimism/optimism/op-service/testlog"
)

func TestSystem(t *testing.T) {
	devEnv, err := env.LoadDevnetFromURL("kt://interop-devnet")
	require.NoError(t, err)
	logger := testlog.Logger(t, log.LevelInfo)
	orch := &Orchestrator{
		env:                &devEnv.Config,
		usePrivatePorts:    false,
		useEagerRPCClients: false,
	}
	setup := &system2.Setup{
		Ctx:          context.Background(),
		Log:          logger,
		T:            t,
		Require:      require.New(t),
		System:       nil,
		Orchestrator: orch,
	}
	setup.System = system2.NewSystem(system2.SystemConfig{
		CommonConfig: setup.CommonConfig(),
	})
	ids, opt := DefaultSystemExt(&devEnv.Config)
	opt(setup)

	for {
		time.Sleep(time.Second * 2)

		l2Net := setup.System.L2Network(ids.L2s[0].L2)
		seq := l2Net.L2CLNode(ids.L2s[0].Nodes[0].CL)
		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
		status, err := seq.RollupAPI().SyncStatus(ctx)
		cancel()
		require.NoError(t, err)
		logger.Info("status", "tip", status.UnsafeL2)

		if status.UnsafeL2.Number > 20 {
			return
		}
	}
}
```

But it appears like the L1 chain config isn't loading. I am not sure what I am doing wrong, or if it's a bug.

sigma committed:
fix: devnet descriptor L1 config

There's more data we need to bubble up from kurtosis.
In the meantime, create a non-empty chain config for L1.

sigma committed:
fix(systemext): fix various issues

Make the following test pass:

```
package systemext

import (
        "context"
        "testing"
        "time"

        "github.com/stretchr/testify/require"

        "github.com/ethereum/go-ethereum/log"

        "github.com/ethereum-optimism/optimism/devnet-sdk/shell/env"
        "github.com/ethereum-optimism/optimism/devnet-sdk/system2"
        "github.com/ethereum-optimism/optimism/op-service/testlog"
)

func TestSystem(t *testing.T) {
        devEnv, err := env.LoadDevnetFromURL("kt://interop-devnet")
        require.NoError(t, err)
        logger := testlog.Logger(t, log.LevelInfo)
        setup := &system2.Setup{
                Ctx:     context.Background(),
                Log:     logger,
                T:       t,
                Require: require.New(t),
                System:  nil,
        }
        setup.System = system2.NewSystem(system2.SystemConfig{
                CommonConfig: setup.CommonConfig(),
        })
        ids, opt := DefaultSystemExt(&devEnv.Config, WithEagerRPCClients())
        opt(setup)

        for {
                time.Sleep(time.Second * 2)

                l2Net := setup.System.L2Network(ids.L2s[0].L2)
                seq := l2Net.L2CLNode(ids.L2s[0].Nodes[0].CL)
                ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
                status, err := seq.RollupAPI().SyncStatus(ctx)
                cancel()
                require.NoError(t, err)
                logger.Info("status", "tip", status.UnsafeL2)

                if status.UnsafeL2.Number > 20 {
                        return
                }
        }
}
```

sigma commented:
Added some fixes to make the following pass (slightly modified from above)


```go
package systemext

import (
        "context"
        "testing"
        "time"

        "github.com/stretchr/testify/require"

        "github.com/ethereum/go-ethereum/log"

        "github.com/ethereum-optimism/optimism/devnet-sdk/shell/env"
        "github.com/ethereum-optimism/optimism/devnet-sdk/system2"
        "github.com/ethereum-optimism/optimism/op-service/testlog"
)

func TestSystem(t *testing.T) {
        devEnv, err := env.LoadDevnetFromURL("kt://interop-devnet")
        require.NoError(t, err)
        logger := testlog.Logger(t, log.LevelInfo)
        setup := &system2.Setup{
                Ctx:     context.Background(),
                Log:     logger,
                T:       t,
                Require: require.New(t),
                System:  nil,
        }
        setup.System = system2.NewSystem(system2.SystemConfig{
                CommonConfig: setup.CommonConfig(),
        })
        ids, opt := DefaultSystemExt(&devEnv.Config, WithEagerRPCClients())
        opt(setup)

        for {
                time.Sleep(time.Second * 2)

                l2Net := setup.System.L2Network(ids.L2s[0].L2)
                seq := l2Net.L2CLNode(ids.L2s[0].Nodes[0].CL)
                ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
                status, err := seq.RollupAPI().SyncStatus(ctx)
                cancel()
                require.NoError(t, err)
                logger.Info("status", "tip", status.UnsafeL2)

                if status.UnsafeL2.Number > 20 {
                        return
                }
        }
}
```

sigma commented:
added a workaround for the L1 chain config. Proper fix will be tracked at #15050

tynes commented on kurtosis-devnet/pkg/kurtosis/kurtosis.go:
```diff
@@ -162,10 +170,17 @@ func (d *KurtosisDeployer) GetEnvironmentInfo(ctx context.Context, spec *spec.En
 	}
 
+	// Get dependency set
+	depsetData, err := d.depsetExtractor.ExtractData(ctx, d.enclave)
```
thoughts on https://github.com/ethereum-optimism/optimism/issues/15048 to solve this?

protolambda replied:
This is the full copy of the dependency set config, not a snapshot.
We can maybe reduce it to a snapshot-getter in the test frontend, but the go in-process backend counterpart needs the full one, to be able to spin up additional supervisor nodes.

And kurtosis is also being extended to support more supervisor nodes, so I think it's useful to have in the underlying kurtosis package.

And then once it's available, the test frontend can present it to the test itself also, like the rollup config and chain configs are, so the test is aware of what's configured.

sigma replied:
I think if we can avoid passing that dependency set around, and have it discovered/computed instead, then it's great to have the tests do just that.

But as long as that's effectively a deployment artifact, I'd rather take it from a static location than communicating with the system under test, as a bug in that part of the system might then prevent unrelated tests from running.

protolambda commented:
See CI go-test error, the L1 ChainID fixup that fixes the full end to end run of kurtosis doesn't work in the unit-tests for some reason.
This PR is good to merge after that's fixed

sigma commented:
> See CI go-test error, the L1 ChainID fixup that fixes the full end to end run of kurtosis doesn't work in the unit-tests for some reason. This PR is good to merge after that's fixed

oops forgot to add a file. done
